src/groups.ts

```diff
@@ -37,31 +44,120 @@ export const selectGroup = (
     selectedGroupIds: { ...appState.selectedGroupIds, [groupId]: true },
     selectedElementIds: {
       ...appState.selectedElementIds,
-      ...Object.fromEntries(
-        elementsInGroup.map((element) => [element.id, true]),
-      ),
-    },
+      ...elementsInGroup,
+    } as AppState["selectedElementIds"],
   };
 };
 
+export const selectGroups = (function () {
```

Comment:
I rewrote this function for the hot path to improve performance - introduced more efficient algorithms and added caching.